BitFaster.Caching/SynchronizedValueFactory.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BitFaster.Caching
+{
+    public class SynchronizedValueFactory<K, V>
+    {
+        private readonly ICache<K, V> cache;
+        private readonly Func<K, V> valueFactory;
+        private readonly Func<K, Task<V>> valueFactoryAsync;
+        private SingletonCache<K, SemaphoreSlim> semaphoreCache = new SingletonCache<K, SemaphoreSlim>();
+
+        public SynchronizedValueFactory(ICache<K, V> cache, Func<K, V> valueFactory, Func<K, Task<V>> valueFactoryAsync)
+        { 
+            this.cache = cache;
+            this.valueFactory = valueFactory;
+            this.valueFactoryAsync = valueFactoryAsync;
+        }
+
+        public SynchronizedValueFactory(ICache<K, V> cache,  Func<K, V> valueFactory)
+        { 
+            this.cache = cache;
+            this.valueFactory = valueFactory;
+            this.valueFactoryAsync = k => Task.FromResult(valueFactory(k));
+        }
+
+        public SynchronizedValueFactory(ICache<K, V> cache, Func<K, Task<V>> valueFactoryAsync)
+        { 
+            this.cache = cache;
+            this.valueFactory = k => valueFactoryAsync(k).GetAwaiter().GetResult();
+            this.valueFactoryAsync = valueFactoryAsync;
+        }
+
+        public V Create(K key)
+        {
+            using (var sema = this.semaphoreCache.Acquire(key, k => new SemaphoreSlim(1, 1)))
+            {
+                sema.Value.Wait();
+
+                try
+                {
+                    return this.cache.GetOrAdd(key, this.valueFactory);
```

Comment:
This is not correct, since in the GetOrAdd method LRU will find it can't add this value, then try to dispose it if it is IDisposable.

This would result in there being a disposed value in the cache. So probably here we should return default(T).